<a href="https://colab.research.google.com/github/pedroantoniolli-ops/Infnet/blob/main/PD_PedroAntoniolli_FundamentosMachineLearningScikitLearn_v1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

- Aluno: Pedro Domingos Antoniolli
- Professor: Ms Icaro Augusto Maccari Zelioli
- Disciplina: Fundamentos de Machine Learning com Scikit Learn - Fev / Mar 2026

Notebook esta compartilhado no Google Colab e no Github

**Parte 1: Libraries necessarias e escolha da Base de Dados**

1. Será utilizada a base de dados do Kaggle de Analise de Churn. Esta disponivel no link
https://www.kaggle.com/datasets/blastchar/telco-customer-churn

**Bibliotecas Necessarias **

In [ ]:
# Célula: imports
import warnings
import numpy as np
import pandas as pd
from google.colab import drive
import matplotlib.pyplot as plt
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.feature_selection import SelectKBest, f_regression
from sklearn.metrics import classification_report, precision_score, r2_score
from sklearn.preprocessing import RobustScaler, PowerTransformer, KBinsDiscretizer, OneHotEncoder, StandardScaler
import seaborn as sns
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import classification_report, ConfusionMatrixDisplay
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Perceptron
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

import warnings
warnings.filterwarnings("ignore")

sns.set(style="whitegrid")
%matplotlib inline


Instalar a library do Kaggle e kagglehub

In [ ]:
# Instalar Kaggle

!pip install kaggle

In [ ]:
# Instalar Kagglehub e importar dataset

!pip install kagglehub
import kagglehub

Importar o dataframe mall-customers do kaggle

In [ ]:
## abrindo o arquivo diretamente do kaggle

## file_path = 'https://www.kaggle.com/blastchar/telco-customer-churn'

path_base_churn = kagglehub.dataset_download("blastchar/telco-customer-churn")

!ls path_base_churn -lh



Using Colab cache for faster access to the 'telco-customer-churn' dataset.
ls: cannot access 'path_base_churn': No such file or directory


criar um dataframe local de nome base_churn

In [ ]:
## criar um dataframe local csv com base no que foi baixado do Kaggle

df_base_churn = pd.read_csv(f"{path_base_churn}/WA_Fn-UseC_-Telco-Customer-Churn.csv")
df_base_churn.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


**Justificativa:**  O dataset Telco Customer Churn é particularmente adequado para realizar análise exploratória e modelagem com scikit-learn porque possui características que refletem problemas reais de negócio e, ao mesmo tempo, apresenta estrutura apropriada para aplicar técnicas clássicas de machine learning supervisionado.

O churn de clientes é um problema crítico em diversos setores, como telecomunicações, bancos, seguros e serviços digitais. Prever quais clientes têm maior probabilidade de cancelar um serviço permite que as empresas adotem estratégias de retenção direcionadas, reduzindo perdas financeiras e melhorando o relacionamento com o cliente.

Por esse motivo, o dataset permite analisar um problema de decisão realista, no qual o modelo de machine learning pode apoiar ações estratégicas, como campanhas de retenção ou ofertas personalizadas.

A variável alvo Churn possui dois possíveis valores:
- cliente cancela o serviço
- cliente permanece ativo

Isso caracteriza um problema de classificação binária supervisionada, um dos cenários mais comuns em machine learning. Esse tipo de problema é diretamente suportado por diversos algoritmos da biblioteca scikit-learn, como:

- Perceptron
- Regressão logística
- Árvores de decisão
- Support Vector Machines
- Random Forest

Assim, o dataset permite comparar diferentes modelos de classificação de forma consistente.

O dataset possui aproximadamente 7.000 registros de clientes.
Esse volume é suficientemente grande para:
- treinar modelos supervisionados
- realizar divisão entre treino e teste
- aplicar validação cruzada
- avaliar a generalização dos modelos

Ao mesmo tempo, não é grande a ponto de exigir infraestrutura computacional complexa, sendo ideal para experimentos em ambiente acadêmico.

O dataset possui mais de 20 variáveis, incluindo:
- características demográficas
- informações contratuais
- comportamento de consumo
- histórico financeiro

Isso permite realizar análise multivariada, explorando relações entre diferentes fatores que influenciam o churn.

Além disso, atende ao requisito de possuir mais de 10 features, permitindo aplicar técnicas como:
- seleção de variáveis
- redução de dimensionalidade (por exemplo, PCA)
- analise de importancia de features

O dataset contém dois tipos principais de variáveis.
Variáveis contínuas, tais como:
- tenure (tempo como cliente),
- MonthlyCharges (valor mensal),
- TotalCharges (valor total pago)

Essas variáveis permitem explorar distribuição, correlações e escalonamento de dados.

Variáveis categóricas, tais como:
- tipo de contrato
- método de pagamento
- serviços adicionais

Essas variáveis exigem técnicas de pré-processamento, como One-Hot Encoding, que são implementadas com ferramentas do scikit-learn.

Essa diversidade torna o dataset ideal para demonstrar pipelines completos de machine learning.

Normalmente, datasets de churn apresentam menos clientes que cancelam do que clientes ativos, cenario que cria desafio importante (conforme citado em aula):

- modelos podem aprender a prever apenas a classe majoritária

Isso permite discutir conceitos importantes como:
- precisão (precision)
- revocação (recall)
- F1-score
- impacto de erros de classificação

Essas métricas fazem parte da avaliação padrão no scikit-learn.

Outro aspecto importante é que muitas features possuem significado claro no domínio de negócio, como:
- tipo de contrato
- valor da mensalidade
- tempo como cliente

Isso permite interpretar resultados de modelos como:
- coeficientes do Perceptron
- regras da árvore de decisão
- importância das variáveis em ensembles

Assim, o dataset facilita conectar os resultados do modelo com explicações compreensíveis para gestores ou analistas de negócio.

Portanto, o dataset de churn de clientes é particularmente adequado para análise exploratória e modelagem com scikit-learn porque combina:
- um problema real de decisão empresarial
- classificação binária bem definida
- número adequado de observações
- múltiplas features relevantes
- mistura de variáveis contínuas e categóricas
- possibilidade de interpretação dos resultados

Essas características tornam o dataset apropriado para explorar todas as etapas de um projeto de machine learning supervisionado, desde a análise exploratória até a comparação de diferentes modelos preditivos.



In [ ]:
# Ignorar todos os warnings

warnings.filterwarnings('ignore')

In [ ]:
# Definir o randon state (definido 19 por ser a data do meu aniversario)

RANDON_STATE = 19

Visualizacao inicial do dataset

In [ ]:
print(base_churn)

/kaggle/input/telco-customer-churn


Pré-processamento (limpeza, tratamento de nulos, scaling)

In [ ]:
# Limpeza e preparacao minima
# Remover "customer ID" (identificador, nao preditivo)
# Converter "TotalCharges" para numerico (algumas linhas chegam com espacos em branco)
# Variavel Alvo = "Churn" => Mapear Yes/No para 1/0

# Create a working copy of df_base_churn
df = df_base_churn.copy()

print("Shape do dataset:", df.shape)
print(df.head())

# remover coluna de ID
df = df.drop(columns=["customerID"])

# converter TotalCharges para número
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")

# remover valores nulos
df = df.dropna()

Shape do dataset: (7043, 21)
   customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV Streami

A variável-alvo (também chamada de target, label ou variável dependente) é a variável que o modelo de machine learning tenta prever a partir das demais variáveis do dataset.

Em um problema supervisionado, o dataset é dividido em duas partes:
- Features (X) → variáveis usadas para fazer a previsão
- Variável-alvo (y) → resultado que queremos prever

Neste caso:

Features
- idade
- renda
- tempo_cliente

Variável-alvo
- cancelou

In [ ]:
# ============================================
# DEFINIÇÃO DA VARIÁVEL-ALVO
# ============================================

df["Churn"] = df["Churn"].map({"Yes":1, "No":0})

X = df.drop(columns=["Churn"])
y = df["Churn"]


In [ ]:
# ============================================
# IDENTIFICAÇÃO DE FEATURES
# ============================================

numeric_features = X.select_dtypes(include=["int64","float64"]).columns
categorical_features = X.select_dtypes(include=["object"]).columns

print("Numéricas:", numeric_features)
print("Categóricas:", categorical_features)



Numéricas: Index(['SeniorCitizen', 'tenure', 'MonthlyCharges', 'TotalCharges'], dtype='object')
Categóricas: Index(['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod'],
      dtype='object')


O pré-processamento de dados é a etapa em que os dados brutos do dataset são limpos, transformados e preparados antes de serem utilizados no treinamento de modelos de machine learning. Essa etapa é essencial porque, na maioria dos casos, os dados reais não estão imediatamente prontos para uso por algoritmos de aprendizado.

No contexto do projeto, o pré-processamento garante que os modelos (como Perceptron, Árvore de Decisão ou Random Forest) consigam interpretar corretamente as variáveis e aprender padrões relevantes.

Muitos algoritmos de machine learning trabalham apenas com valores numéricos. Entretanto, datasets reais frequentemente contêm variáveis categóricas.

Exemplo do dataset de churn:

Modelos como Perceptron não conseguem trabalhar diretamente com texto. Por isso, utilizamos técnicas como codificação categórica (One-Hot Encoding) para transformar categorias em variáveis numéricas.

Outro ponto: algoritmos de machine learning geralmente não conseguem lidar com valores nulos, portanto é necessário:
- remover registros incompletos
- substituir valores faltantes
- converter tipos incorretos

No dataset de churn, por exemplo, a variável TotalCharges precisa ser convertida para número e os valores inválidos devem ser tratados.

Alem disso, datasets reais frequentemente possuem valores faltantes.

Algoritmos de machine learning geralmente não conseguem lidar com valores nulos, portanto é necessário:
- remover registros incompletos
- substituir valores faltantes
- converter tipos incorretos

No dataset de churn, por exemplo, a variável TotalCharges precisa ser convertida para número e os valores inválidos devem ser tratados.

Outro ponto a ser considerado seria a normalizacao, ou seja, alguns algoritmos são sensíveis à escala das variáveis.

Sem normalização, variáveis com valores maiores podem dominar o aprendizado do modelo.

Por isso utilizamos técnicas como StandardScaler, que padroniza os dados para média 0 e desvio padrão 1.

Isso é especialmente importante para modelos lineares como:
- Perceptron
- Regressão logística
- SVM

O pre processamento ajuda a:
- remover ruído
- corrigir inconsistências
- padronizar formatos

Dessa forma, dados bem preparados permitem que o modelo:
- aprenda padrões mais claros
- evite erros de interpretação
- produza previsões mais confiáveis

Sem pré-processamento, mesmo modelos avançados podem apresentar desempenho ruim.

Adicionalmente, ao incluir o pré-processamento em um pipeline, todo o processo de transformação dos dados fica documentado e reproduzível.

Isso é essencial em projetos profissionais e acadêmicos, pois garante que o modelo possa ser treinado novamente nas mesmas condições.

Em resumo, o pré-processamento é uma etapa fundamental em machine learning porque transforma dados brutos em dados adequados para treinamento de modelos. Ele permite corrigir inconsistências, lidar com valores ausentes, transformar variáveis categóricas em numéricas e ajustar a escala das variáveis.

Sem essa etapa, os algoritmos podem não funcionar corretamente ou produzir resultados pouco confiáveis.


In [ ]:
# ============================================
# PRÉ-PROCESSAMENTO
# ============================================

numeric_transformer = Pipeline(
    steps=[
        ("scaler", StandardScaler())
    ]
)

categorical_transformer = Pipeline(
    steps=[
        ("encoder", OneHotEncoder(handle_unknown="ignore"))
    ]
)

preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_transformer, numeric_features),
        ("cat", categorical_transformer, categorical_features)
    ]
)

Dividir o dataset em conjunto de treino (training set) e conjunto de teste (test set) é uma prática fundamental em machine learning supervisionado, pois permite avaliar corretamente a capacidade de generalização do modelo.

Em outras palavras, queremos verificar se o modelo funciona bem em dados que ele nunca viu antes.

Por exemplo, para evitar avaliacao otimista do modelo.

Se treinarmos e avaliarmos o modelo usando os mesmos dados, o resultado será artificialmente bom.

Isso ocorre porque o modelo pode memorizar padrões específicos do dataset, ao invés de aprender relações gerais.

Exemplo:
- modelo treinado com 1000 registros
- testado nos mesmos 1000 registros

Resultado possível:
- Accuracy = 99%

Mas ao aplicar em novos dados:
- Accuracy pode cair para 60% ou menos

Esse problema é chamado de overfitting.

Alem disso, sse torna adequado para simular dados futuros (nunca vistos)

O conjunto de teste representa novas observações que apareceriam no mundo real.

Processo:
1. Treinar o modelo usando dados de treino
2. Aplicar o modelo em dados de teste
3. Medir o desempenho com métricas como:

- Accuracy
- Precision
- Recall
- F1-score

Assim avaliamos a capacidade real de generalização do modelo.

Pipeline:

No contexto de machine learning com scikit-learn, um pipeline é uma estrutura que permite encadear várias etapas de processamento e modelagem em um único fluxo organizado.

Ou seja, ele define uma sequência automática de transformações nos dados até chegar ao modelo final.

Um pipeline conecta etapas como:
- pré-processamento dos dados
- transformação das features
- treinamento do modelo

Tudo executado de forma sequencial.
Assim, o pipeline garante que todas as etapas sejam executadas na mesma ordem sempre que o modelo for treinado ou usado para prever novos dados.

In [ ]:
# ============================================
# DIVISÃO TREINO / TESTE
# ============================================

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=RANDON_STATE
)

A função de avaliação é o conjunto de métricas usadas para medir o desempenho de um modelo de machine learning após o treinamento. Ela permite verificar quão bem o modelo está fazendo previsões ao comparar os valores previstos com os valores reais da variável-alvo.

Neste projeto, a função de avaliação serve para analisar a qualidade dos modelos treinados, como:
- Perceptron (baseline)
- Árvore de decisão
- Random Forest ou SVM

E comparar qual deles apresenta melhor desempenho.

Após treinar um modelo, fazemos previsões:
- modelo → gera previsões

Essas previsões são comparadas com os valores reais do dataset de teste.
- valor real vs valor previsto

A função de avaliação calcula métricas que indicam o nível de acerto e erro do modelo.

In [ ]:
# ============================================
# FUNÇÃO DE AVALIAÇÃO
# ============================================

def evaluate_model(model, X_test, y_test):

    y_pred = model.predict(X_test)

    print("Accuracy:", accuracy_score(y_test, y_pred))
    print("Precision:", precision_score(y_test, y_pred))
    print("Recall:", recall_score(y_test, y_pred))
    print("F1-score:", f1_score(y_test, y_pred))

    print("\nRelatório completo")
    print(classification_report(y_test, y_pred))


O Perceptron no scikit-learn é um classificador linear supervisionado utilizado principalmente para problemas de classificação binária. Ele é um dos algoritmos mais clássicos da área de redes neurais artificiais e também pode ser visto como um modelo linear simples que aprende um hiperplano de separação entre classes.

O Perceptron aprende uma função linear que combina as features do dataset para produzir uma decisão de classe.

Se houver apenas duas variáveis, o modelo aprende uma linha separando as classes.

Se houver três variáveis, aprende um plano.

Com muitas variáveis, temos um hiperplano em espaço multidimensional.

O objetivo do treinamento é ajustar os pesos 𝑤 para minimizar erros de classificação.

O algoritmo segue um processo iterativo:

1. Inicializa os pesos aleatoriamente.
2. Analisa uma observação.
3. Faz uma previsão.
4. Se a previsão estiver errada, ajusta os pesos.

Algumas vantagens do Perceptrom
- Simples
- Rapido
- Altamente interpretavel
- bom como baseline de modelos lineares

In [ ]:
# ============================================================
# MODELO BASELINE — PERCEPTRON
# ============================================================

print("\n==============================")
print("MODELO BASELINE — PERCEPTRON")
print("==============================")

perceptron_pipeline = Pipeline(
    steps=[
        ("preprocess", preprocessor),
        ("model", Perceptron(random_state=RANDON_STATE))
    ]
)

perceptron_pipeline.fit(X_train, y_train)

evaluate_model(perceptron_pipeline, X_test, y_test)

# coeficientes do perceptron
model = perceptron_pipeline.named_steps["model"]

print("\nCoeficientes do modelo:")
print(model.coef_)

print("\nBias (intercepto):")
print(model.intercept_)




MODELO BASELINE — PERCEPTRON
Accuracy: 0.6936744847192609
Precision: 0.45640074211502785
Recall: 0.640625
F1-score: 0.5330444203683641

Relatório completo
              precision    recall  f1-score   support

           0       0.84      0.71      0.77      1023
           1       0.46      0.64      0.53       384

    accuracy                           0.69      1407
   macro avg       0.65      0.68      0.65      1407
weighted avg       0.74      0.69      0.71      1407


Coeficientes do modelo:
[[-1.86572042 -8.76494608 -4.13313174 -1.65335929 -2.          0.
  -1.         -1.          2.         -4.         -3.          1.
   2.         -3.         -1.         -3.          2.         -1.
   0.         -1.         -1.         -2.         -1.          1.
   2.         -1.         -3.          0.         -1.         -1.
  -2.         -1.          1.         -3.         -1.          2.
   4.          0.         -6.          0.         -2.          4.
  -1.          1.         -6. 

Arvore de decisao:
A árvore de decisão é um modelo de machine learning supervisionado que realiza classificações ou previsões por meio de uma estrutura hierárquica de decisões, semelhante a um fluxograma. Esse modelo divide os dados em subconjuntos cada vez mais específicos, com base nos valores das variáveis (features), até chegar a uma decisão final.

Ela é muito utilizada porque é fácil de interpretar, explica claramente as regras de decisão e consegue capturar relações não lineares entre variáveis.

Uma árvore de decisão possui três elementos principais:

1. Nó raiz (root node)

É o primeiro ponto da árvore, onde ocorre a primeira divisão dos dados com base em uma variável considerada mais relevante para a previsão.

2. Nós internos (decision nodes)

São pontos intermediários da árvore que representam novas regras de decisão baseadas em outras variáveis.

Cada decisão divide o conjunto de dados em grupos menores.

3. Nós folha (leaf nodes)

São os resultados finais da classificação.

Para que a arvore possa aprender, durante o treinamento, o algoritmo procura dividir os dados de forma que as classes fiquem o mais separadas possível.

Ele escolhe a divisão que melhor separa os dados usando critérios matemáticos como:

- Gini Impurity: Mede o nível de mistura das classes em um nó. Quanto menor o valor, mais puro é o nó.

- Entropy (informação): Baseado na teoria da informação, mede o grau de desordem dos dados.

O objetivo é criar nós onde a maioria das observações pertença à mesma classe.

Alguns parâmetros controlam a complexidade da árvore.

- max_depth:	profundidade máxima da árvore
- min_samples_split	mínimo de observações para dividir um nó
- min_samples_leaf	mínimo de observações em um nó folha
- criterion	critério de divisão (gini ou entropy)

Esses parâmetros ajudam a evitar overfitting.

Vantagens da árvore de decisão
- fácil de interpretar
- gera regras claras
- captura relações não lineares
- funciona bem com dados categóricos e numéricos

Limitacoes
- Overfitting: Árvores muito profundas podem memorizar os dados de treino.
- Instabilidade: Pequenas mudanças nos dados podem gerar árvores diferentes.
- Menor desempenho isoladamente; Em muitos casos, modelos baseados em ensemble (Random Forest, Gradient Boosting) apresentam melhor desempenho.

Neste trabalho, a árvore de decisão será usada para:
1. comparar com o modelo baseline (Perceptron)
2. avaliar se um modelo não linear melhora a classificação
3. interpretar as regras aprendidas pelo modelo

Depois sera aplidado:
- validação cruzada
- busca de hiperparâmetros (GridSearch)
Para melhorar a árvore.

Em resumo, a árvore de decisão é um modelo de machine learning que aprende regras de decisão em forma de árvore para classificar dados. Ela divide o espaço das features em regiões menores, criando regras que levam da raiz até uma decisão final.

Esse modelo é particularmente útil por ser interpretável e capaz de capturar relações complexas entre variáveis.


In [ ]:
# ============================================================
# MODELO 2 — ÁRVORE DE DECISÃO
# ============================================================

print("\n==============================")
print("ÁRVORE DE DECISÃO")
print("==============================")

tree_pipeline = Pipeline(
    steps=[
        ("preprocess", preprocessor),
        ("model", DecisionTreeClassifier(random_state=RANDON_STATE))
    ]
)

tree_pipeline.fit(X_train, y_train)

evaluate_model(tree_pipeline, X_test, y_test)


ÁRVORE DE DECISÃO
Accuracy: 0.7320540156361052
Precision: 0.5087281795511222
Recall: 0.53125
F1-score: 0.5197452229299363

Relatório completo
              precision    recall  f1-score   support

           0       0.82      0.81      0.81      1023
           1       0.51      0.53      0.52       384

    accuracy                           0.73      1407
   macro avg       0.66      0.67      0.67      1407
weighted avg       0.74      0.73      0.73      1407



A validação cruzada (cross-validation) combinada com Grid Search é uma técnica usada em machine learning para avaliar modelos de forma mais confiável e encontrar os melhores hiperparâmetros.

Essa etapa é muito importante porque ajuda a garantir que o modelo generalize bem para novos dados, evitando problemas como overfitting.

A validação cruzada consiste em dividir o conjunto de dados de treinamento em vários subconjuntos chamados folds.

Um método comum é o k-fold cross-validation.

In [ ]:
# ============================================================
# VALIDAÇÃO CRUZADA + GRID SEARCH
# ============================================================

print("\n==============================")
print("OTIMIZAÇÃO DA ÁRVORE")
print("==============================")

param_grid = {
    "model__max_depth": [3,5,10,None],
    "model__min_samples_split": [2,5,10],
    "model__min_samples_leaf": [1,2,5]
}

grid = GridSearchCV(
    tree_pipeline,
    param_grid,
    cv=5,
    scoring="f1",
    n_jobs=-1
)

grid.fit(X_train, y_train)

print("Melhores parâmetros:")
print(grid.best_params_)

best_tree = grid.best_estimator_

evaluate_model(best_tree, X_test, y_test)



OTIMIZAÇÃO DA ÁRVORE
Melhores parâmetros:
{'model__max_depth': 5, 'model__min_samples_leaf': 5, 'model__min_samples_split': 2}
Accuracy: 0.7967306325515281
Precision: 0.6346153846153846
Recall: 0.6015625
F1-score: 0.6176470588235294

Relatório completo
              precision    recall  f1-score   support

           0       0.85      0.87      0.86      1023
           1       0.63      0.60      0.62       384

    accuracy                           0.80      1407
   macro avg       0.74      0.74      0.74      1407
weighted avg       0.79      0.80      0.79      1407



A Random Forest (Floresta Aleatória) é um algoritmo de machine learning supervisionado baseado em ensemble learning, ou seja, ele combina vários modelos simples para produzir uma previsão mais robusta e precisa.

Nesse caso, o modelo combina múltiplas árvores de decisão, formando uma “floresta”. Cada árvore faz uma previsão e o resultado final é obtido por votação entre as árvores.

Uma Random Forest funciona da seguinte maneira:
1. O algoritmo cria várias árvores de decisão.
2. Cada árvore é treinada com uma amostra aleatória dos dados.
3. Cada árvore utiliza um subconjunto aleatório de features.
4. As previsões de todas as árvores são combinadas.

Para classificação, o resultado final é a classe mais votada.

Por que usar varias arvores?

Uma única árvore de decisão pode ser instável e propensa a overfitting.

A Random Forest reduz esse problema porque:
- cada árvore vê dados diferentes
- cada árvore usa features diferentes
- o resultado final é a média das decisões

Isso torna o modelo:
- mais estável
- mais robusto
- mais preciso

In [ ]:

# ============================================================
# MODELO AVANÇADO — RANDOM FOREST
# ============================================================

print("\n==============================")
print("RANDOM FOREST")
print("==============================")

rf_pipeline = Pipeline(
    steps=[
        ("preprocess", preprocessor),
        ("model", RandomForestClassifier(random_state=RANDON_STATE))
    ]
)

rf_params = {
    "model__n_estimators": [100,200],
    "model__max_depth": [None,10,20],
    "model__min_samples_split": [2,5]
}

rf_grid = GridSearchCV(
    rf_pipeline,
    rf_params,
    cv=5,
    scoring="f1",
    n_jobs=-1
)

rf_grid.fit(X_train, y_train)

print("Melhores parâmetros Random Forest:")
print(rf_grid.best_params_)

best_rf = rf_grid.best_estimator_

evaluate_model(best_rf, X_test, y_test)



RANDOM FOREST
Melhores parâmetros Random Forest:
{'model__max_depth': 10, 'model__min_samples_split': 5, 'model__n_estimators': 200}
Accuracy: 0.7953091684434968
Precision: 0.6655172413793103
Recall: 0.5026041666666666
F1-score: 0.5727002967359051

Relatório completo
              precision    recall  f1-score   support

           0       0.83      0.91      0.87      1023
           1       0.67      0.50      0.57       384

    accuracy                           0.80      1407
   macro avg       0.75      0.70      0.72      1407
weighted avg       0.78      0.80      0.79      1407



Comparar diferentes modelos de machine learning é uma etapa fundamental porque permite identificar qual algoritmo resolve melhor o problema analisado, considerando desempenho, capacidade de generalização, interpretabilidade e custo computacional.

No contexto deste projeto, a comparação entre modelos como Perceptron, Árvore de Decisão e Random Forest ajuda a entender como diferentes abordagens de aprendizado lidam com o mesmo conjunto de dados.

Cada algoritmo possui características diferentes e pode ter desempenhos distintos dependendo da natureza dos dados.

Ao comparar modelos utilizando métricas como:
- Accuracy
- Precision
- Recall
- F1-score

é possível identificar qual modelo produz previsões mais corretas.

In [ ]:


# ============================================================
# COMPARAÇÃO DE MODELOS
# ============================================================

print("\n==============================")
print("COMPARAÇÃO FINAL")
print("==============================")

models = {
    "Perceptron": perceptron_pipeline,
    "Decision Tree": tree_pipeline,
    "Optimized Tree": best_tree,
    "Random Forest": best_rf
}

for name, model in models.items():

    scores = cross_val_score(
        model,
        X,
        y,
        cv=5,
        scoring="f1"
    )

    print(name)
    print("F1 médio:", scores.mean())
    print("Desvio padrão:", scores.std())
    print()


COMPARAÇÃO FINAL
Perceptron
F1 médio: 0.44598002703959594
Desvio padrão: 0.16250570470799391

Decision Tree
F1 médio: 0.4889953165258838
Desvio padrão: 0.021210218251634518

Optimized Tree
F1 médio: 0.6002513140361235
Desvio padrão: 0.017098391672827857

Random Forest
F1 médio: 0.5783903022152924
Desvio padrão: 0.019506329543831354



**Interpretacao dos Resultados**

**Perceptron**

O F1-score médio representa o desempenho médio do modelo ao longo dos folds da validação cruzada.

O F1-score é uma métrica que combina:
- Precision → qualidade das previsões positivas
- Recall → capacidade de encontrar casos positivos

Ele é muito utilizado quando:
- há desbalanceamento de classes
- queremos equilíbrio entre falsos positivos e falsos negativos

No caso do churn, isso é importante porque queremos identificar clientes que irão cancelar sem gerar muitos alarmes falsos.

O desvio padrão mede a variação do desempenho entre os folds da validação cruzada.

Interpretação:
- baixo desvio padrão → modelo estável
- alto desvio padrão → modelo instável

Isso indica consistência do modelo ao generalizar para diferentes subconjuntos dos dados.

- F1 médio: 0.446
- Desvio padrão: 0.163

O Perceptron apresentou o pior desempenho entre os modelos avaliados. Isso acontece porque:
- ele é um classificador linear
- o problema de churn provavelmente possui relações não lineares entre variáveis

Outro ponto importante:

O desvio padrão é alto (0.16), indicando que o desempenho varia bastante entre os folds.

Isso sugere que:
- o modelo é instável
- tem dificuldade em generalizar para diferentes subconjuntos de dados.

Dessa forma, o Perceptron funciona apenas como baseline, confirmando que modelos lineares simples não são suficientes para capturar os padrões do problema.

Arvore de Decisao

- F1 médio: 0.489
- Desvio padrão: 0.021

A árvore de decisão apresenta melhor desempenho que o Perceptron.

Isso ocorre porque:
- árvores conseguem capturar relações não lineares
- conseguem modelar interações entre variáveis

Além disso:

O desvio padrão é baixo, indicando que o modelo é estável entre os folds, porém, o desempenho ainda é relativamente moderado.

Possível motivo:
- árvore simples pode não capturar toda a complexidade do problema.

Arvore Otimizada

- F1 médio: 0.600
- Desvio padrão: 0.017

Este foi o melhor modelo entre todos.

A melhoria ocorreu após:
- otimização de hiperparâmetros
- controle de complexidade da árvore

Isso mostra a importância de:
- regularização
- busca sistemática de hiperparâmetros

O desvio padrão é muito baixo, indicando que o modelo é:
- estável
- robusto

A otimização permitiu melhorar significativamente o desempenho do modelo sem aumentar excessivamente sua complexidade.

Random Forest

- F1 médio: 0.578
- Desvio padrão: 0.019

A Random Forest apresenta bom desempenho, superior à árvore simples, mas ligeiramente inferior à árvore otimizada.
- Isso pode acontecer porque:
- o conjunto de dados não é muito grande
- a árvore otimizada já capturou bem os padrões

O desvio padrão também é baixo, indicando boa estabilidade do modelo.

Random Forest oferece bom equilíbrio entre desempenho e generalização, mas neste caso específico não superou o modelo de árvore otimizada.

Comparacao entre os modelos
- Perceptron - FI 0.446 - desempenho baixo
- Arvore de decisao - FI 0.489 - Melhoria com nao linearidade
- Arvore otimizada - FI 0.600 - Melhor desempenho
- Random Forest - FI 0.578 - Modelo robusto


**REFERENCIAS**

GALLATIN, K.; ALBON, C. Machine Learning with Python Cookbook, 2nd Edition. Disponivel em O'Reilly. Acesso em 03 de Marco de 2026.

GERON, A. Hands-On Machine Learning with Scikit-Learn and PyTorch. Disponivel em O'Reilly. Acesso em 03 de Marco de 2026.

ZELIOLI, I. A. M. Fundamentos de Machine Learning com Scikit Learning. MIT Inteligencia Artificial. Notas de Aula. 2026.

SCIKIT-LEARN (biblioteca do software de programação Python) https://scikitlearn.org/stable/auto_examples/cluster/plot_kmeans_silhouette_analysis.html. Acesso em: 07 de dezembro de 2025.

SERRANO, L. Grokking Machine Learning. Disponivel em O'Reilly. Acesso em 03 de Marco de 2026.

SUKUP, J. scikit-learn Cookbook - Third Edition. Disponivel em O'Reilly. Acesso em 03 de Marco de 2026.



